Server-Driven Paging (#35)
SAP/cloud-sdk: docs/java/features/odata/type-safe-client-odata-v2.mdx

</Tabs>

## Server-Driven Paging

> OData service operators may decide to enable server-driven pagination to limit the amount of data that is fetched and sent over the network to the client, in case generic query options yield huge amounts of data. By splitting the result-set into sequential pages of entities, the data can be requested incrementally. This reduces initial network load and improves overall response times. If the payload of an OData response contains a link including a `$skiptoken`, then it indicates a next page to the result-set.

Reviewer: I would point out more clearly that this is specific to reading a collection of entities.

Author: Added introduction sentence.

> OData service operators may decide to enable server-driven pagination to limit the amount of data that is fetched and sent over the network to the client, in case generic query options yield huge amounts of data. By splitting the result-set into sequential pages of entities, the data can be requested incrementally. This reduces initial network load and improves overall response times. If the payload of an OData response contains a link including a `$skiptoken`, then it indicates a next page to the result-set.

Author: ```suggestion
Pagination describes the practice of splitting a collection of entities from reading requests into one or many pages. The paging behavior is determined by both server and client.

OData service operators may decide to enable server-driven pagination to limit the amount of data that is fetched and sent over the network to the client, in case generic query options yield huge amounts of data. By splitting the result-set into sequential pages of entities, the data can be requested incrementally. This reduces initial network load and improves overall response times. If the payload of an OData response contains a link including a `$skiptoken`, then it indicates a next page to the result-set. The iterable result-set is a consistent snapshot of the data, it can not change between reading individual pages.
```

> In comparison the OData consumer may use query options `$top` and `$skip` (client-driven paging) to read partial data from the result-set. But relying on a consistent state while browsing through the data can be problematic. Between individual requests another user could delete or add an item. This would result in an inconsistent aggregation of data.

Reviewer: I would be cautious with mentioning this. Because tchnically top and skip are not pagination because you are not reading a consistent snapshot of the data. The data can change between reading individual "pages". Also it's not performance optimised.

Author: > Because technically top and skip are not pagination because you are not reading a consistent snapshot of the data. 

It's referenced as _client-driven pagination_ in different documentations of OData protocol as well. Also it makes sense to mention $top and $skip to exactly tell the difference towards _server-driven pagination_.

> the data can change between reading individual "pages".

I hope I outlined this challenge already in the original paragraph. I'll add a sentence to the paragraph earlier.

> Also it's not performance optimized.

I wouldn't call either pagination technique as performance optimized (by default). Let's rather be careful with those promises :)

Reviewer: Ah OK I wasn't aware this is mentioned in the documentation. And you are right how this is implemented in the end depends on the service.

> By default the SAP Cloud SDK automatically resolves all pages of a result-set if server-driven paging is encountered. For the API consumer it is not necessary to parse the next link and instantiate follow-up requests to aggregate the full result-set. In case memory efficiency and response time is the priority, then the advanced API provides additional means to manually iterate through the internal pages. While accessing the following methods, the internal HTTP requests are executed lazily:

Reviewer: I would include a short code snippet here of the "standard" `getAllXyz()` usage

```java
Iterable<List<BusinessPartner>> iterablePagesOfEntities = service
  .getAllBusinessPartner()
  .iteratingPages()
  .executeRequest( destination );

Iterable<BusinessPartner> iterableEntities = service
  .getAllBusinessPartner()
  .iteratingEntities()
  .executeRequest(destination);
  
Stream<BusinessPartner> streamingEntities = service
  .getAllBusinessPartner()
  .streamingEntities()
  .executeRequest(destination);
```

> The request builder allows for setting the optional parameter for preferred page size. Please note that the OData service may not be obliged to respect this setting.

Reviewer: ```suggestion
The request builder allows for setting the optional parameter for preferred page size. Please note that the OData service is not obliged to respect this setting.
```
